File: scripts/eval/pip-containment.py
#!/usr/bin/env python3
"""PIP-containment metric (coordinate-first plan, #273).

Reads the `--out-resolved` dump from oa-resolver-eval.ts (per row: gold OA lat/lon + the neural-resolved
locality's WOF id + the old name-match flag) and tests the NON-GAMEABLE truth: does the gold point lie
INSIDE the polygon of the resolved WOF locality? This is name-surface-independent — it rewards a
geographically-correct resolve even when WOF's canonical name ("Plauen") differs from OA's gold
("Plauen Vogtl"). Compares containment-accuracy vs the old name-match on the SAME rows.

Usage: python3 scripts/eval/pip-containment.py /tmp/resolved-de.json
"""
import json, sys, glob, os, collections

WOF_REPOS = "/mnt/playpen/mailwoman-data/wof/repos"
ADMIN_ROOTS = sorted(glob.glob(f"{WOF_REPOS}/whosonfirst-data/whosonfirst-data-admin-*/data")) + \
              [f"{WOF_REPOS}/whosonfirst-data-admin-us/data"]

_geom_cache = {}

def geom_for_id(wof_id):
    if wof_id in _geom_cache:
        return _geom_cache[wof_id]
    s = str(int(wof_id))
    # WOF path: split the id into 3-char chunks (last chunk is the remainder).
    chunks, i = [], 0
    while i < len(s):
        chunks.append(s[i:i+3]); i += 3
    rel = "/".join(chunks) + f"/{s}.geojson"
    geom = None
    for root in ADMIN_ROOTS:
        fp = os.path.join(root, rel)
        if os.path.exists(fp):
            try:
                geom = json.load(open(fp)).get("geometry")
            except Exception:
                geom = None
            break
    _geom_cache[wof_id] = geom
    return geom

def in_ring(x, y, ring):
    inside, n, j = False, len(ring), len(ring) - 1
    for i in range(n):
        xi, yi = ring[i][0], ring[i][1]
        xj, yj = ring[j][0], ring[j][1]
        if ((yi > y) != (yj > y)) and (x < (xj - xi) * (y - yi) / (yj - yi) + xi):
            inside = not inside
        j = i
    return inside

def in_polygon(x, y, poly):  # poly = [outer, hole1, ...] — even-odd handles holes
    c = False
    for ring in poly:
        if in_ring(x, y, ring):
            c = not c
    return c

def contains(geom, lon, lat):
    if not geom:
        return None  # no polygon available
    t = geom.get("type")
    if t == "Polygon":
        return in_polygon(lon, lat, geom["coordinates"])
    if t == "MultiPolygon":
        return any(in_polygon(lon, lat, p) for p in geom["coordinates"])
    return None  # Point geometry etc. — can't contain

# --- arg parsing: <resolved.json> [--label NAME] [--json OUT] ---------------
args = sys.argv[1:]
src = None
label_arg = None
json_out = None
i = 0
while i < len(args):
    a = args[i]
    if a == "--label":
        label_arg = args[i + 1]; i += 2
    elif a == "--json":
        json_out = args[i + 1]; i += 2
    else:
        src = a; i += 1
if not src:
    print("usage: pip-containment.py <resolved.json> [--label NAME] [--json OUT]", file=sys.stderr)
    sys.exit(2)

rows = json.load(open(src))
overall = collections.Counter()
by_state = collections.defaultdict(collections.Counter)
artifact_examples = []
no_poly = 0
for r in rows:
    st = r.get("state") or "??"
    overall["n"] += 1; by_state[st]["n"] += 1
    name_ok = bool(r.get("nameMatch"))
    if name_ok:
        overall["name"] += 1; by_state[st]["name"] += 1
    lid = r.get("neuralLocId")
    contained = contains(geom_for_id(lid), r["lon"], r["lat"]) if lid else None
    if contained is not None:  # a polygon existed and was tested (True or False)
        overall["poly"] += 1; by_state[st]["poly"] += 1
    elif lid:
        no_poly += 1
    if contained:
        overall["pip"] += 1; by_state[st]["pip"] += 1
        if not name_ok and len(artifact_examples) < 12:
            artifact_examples.append(f'  "{r["input"]}"  gold="{r.get("expectedLoc")}"  resolved="{r.get("neuralLoc")}"')

def pct(num, den):
    return f"{100*num/den:.1f}%" if den else "—"

def line(label, c):
    n = c["n"]
    if not n:
        return f"  {label}: n=0"
    # PIP-containment is reported two ways: over ALL rows (strict) and over rows
    # that HAVE a polygon (coverage-adjusted), since WOF point-geometry localities
    # can never PIP-contain and would otherwise count as silent failures.
    return (
        f"  {label:<10} n={n:<5} name-match={pct(c['name'],n):<7} "
        f"PIP-containment={pct(c['pip'],n):<7} delta={100*(c['pip']-c['name'])/n:+.1f}pp  "
        f"PIP/poly={pct(c['pip'],c['poly']):<7} poly-cov={pct(c['poly'],n)}"
    )

print(f"\n=== PIP-containment vs name-match ({src}{' · '+label_arg if label_arg else ''}) ===")
print(line("OVERALL", overall))
for st in sorted(by_state):
    print(line(st, by_state[st]))
print(f"\n  rows resolved-but-polygon-missing: {no_poly}")
print(f"\nMETRIC-ARTIFACT cases (name-match FAILED but gold point IS inside the resolved locality):")
for e in artifact_examples:
    print(e)

if json_out:
    n = overall["n"]
    summary = {
        "label": label_arg,
        "source": src,
        "n": n,
        "name_match": overall["name"] / n if n else None,
        "pip_all": overall["pip"] / n if n else None,
        "pip_poly": overall["pip"] / overall["poly"] if overall["poly"] else None,
        "poly_coverage": overall["poly"] / n if n else None,
        "no_polygon": no_poly,
    }
    with open(json_out, "w") as f:
        json.dump(summary, f, indent=2)
    print(f"\nwrote summary → {json_out}", file=sys.stderr)

File: scripts/eval/pip-containment.ts
/**
 * @copyright Sister Software
 * @license AGPL-3.0
 * @author Teffen Ellis, et al.
 *
 *   PIP-containment metric (coordinate-first plan, #273).
 *
 *   Reads the `--out-resolved` dump from oa-resolver-eval.ts (per row: gold OA lat/lon + the
 *   neural-resolved locality's WOF id + the old name-match flag) and tests the NON-GAMEABLE truth:
 *   does the gold point lie INSIDE the polygon of the resolved WOF locality? This is
 *   name-surface-independent — it rewards a geographically-correct resolve even when WOF's
 *   canonical name ("Plauen") differs from OA's gold ("Plauen Vogtl"). Compares
 *   containment-accuracy vs the old name-match on the SAME rows.
 *
 *   Ported faithfully from scripts/eval/pip-containment.py (pure JSON + filesystem geojson, no
 *   numpy).
 *
 *   Usage: node scripts/eval/pip-containment.ts <resolved.json> [--label
 *   NAME] [--json OUT]
 */

import { dataRootPath } from "@mailwoman/core/data-root"
import { globPaths, readLocalJSONFile } from "@mailwoman/core/fs/readers"
import { writeLocalJSONFile } from "@mailwoman/core/fs/writers"
import { pyFixed } from "@mailwoman/core/numeric"
import { readWOFFeature } from "@mailwoman/core/resources/whosonfirst"
import { runIfScript } from "@mailwoman/core/scripting"
import { parseArguments } from "@mailwoman/core/scripting/arguments"
import { geometryContains, type GeometryLiteral } from "@mailwoman/spatial"

/**
 * Artifact examples collected before the list is truncated.
 */
const MAX_LISTED_ARTIFACTS = 12

const WOF_REPOS = dataRootPath("wof", "repos")

async function adminRoots(): Promise<string[]> {
	let matched: string[]

	try {
		matched = await globPaths(`${WOF_REPOS}/whosonfirst-data/whosonfirst-data-admin-*/data`)
	} catch {
		matched = []
	}

	matched.sort()

	return [...matched, `${WOF_REPOS}/whosonfirst-data-admin-us/data`]
}

const ADMIN_ROOTS = await adminRoots()

const geomCache = new Map<number, GeometryLiteral | null>()

async function geomForID(wofID: number): Promise<GeometryLiteral | null> {
	if (geomCache.has(wofID)) return geomCache.get(wofID)!

	const feature = await readWOFFeature(Math.trunc(wofID), ADMIN_ROOTS)
	const geom = feature?.geometry ?? null

	geomCache.set(wofID, geom)

	return geom
}

type Counter = Record<string, number>

function inc(c: Counter, k: string): void {
	c[k] = (c[k] ?? 0) + 1
}

function get(c: Counter, k: string): number {
	return c[k] ?? 0
}

/**
 * Python `f"{x:+.1f}"` — fixed precision with an always-present sign.
 */
function pySigned(x: number, d: number): string {
	const s = pyFixed(x, d)

	return s.startsWith("-") ? s : "+" + s
}

function padL(s: string, w: number): string {
	return s.padEnd(w)
}

function pyStr(v: unknown): string {
	return v === undefined || v === null ? "None" : String(v)
}

function pct(num: number, den: number): string {
	// Local rather than `formatPercent`: this port renders with `pyFixed` (Python's round-half-even),
	// and `toFixed` rounds half-away — the bytes must match the retired .py report.
	return den ? `${pyFixed((100 * num) / den, 1)}%` : "—"
}

function line(label: string, c: Counter): string {
	const n = get(c, "n")

	if (!n) return `  ${label}: n=0`

	// PIP-containment is reported two ways: over ALL rows (strict) and over rows
	// that HAVE a polygon (coverage-adjusted), since WOF point-geometry localities
	// can never PIP-contain and would otherwise count as silent failures.
	return (
		`  ${padL(label, 10)} n=${padL(String(n), 5)} name-match=${padL(pct(get(c, "name"), n), 7)} ` +
		`PIP-containment=${padL(pct(get(c, "pip"), n), 7)} delta=${pySigned((100 * (get(c, "pip") - get(c, "name"))) / n, 1)}pp  ` +
		`PIP/poly=${padL(pct(get(c, "pip"), get(c, "poly")), 7)} poly-cov=${pct(get(c, "poly"), n)}`
	)
}

interface ResolvedRow {
	state?: string | null
	nameMatch?: unknown
	neuralLocID?: number | null
	lon: number
	lat: number
	input?: string
	expectedLoc?: unknown
	neuralLoc?: unknown
}

async function main(): Promise<number> {
	const { values, positionals } = parseArguments({
		options: { label: { type: "string" }, json: { type: "string" } },
		strict: false,
		allowPositionals: true,
	})

	const src: string | null = positionals[0] ?? null
	const labelArg: string | null = (values.label as string | undefined) ?? null
	const jsonOut: string | null = (values.json as string | undefined) ?? null

	if (!src) {
		console.error("usage: pip-containment.ts <resolved.json> [--label NAME] [--json OUT]")

		return 2
	}

	const rows = await readLocalJSONFile<ResolvedRow[]>(src)
	const overall: Counter = {}
	const byState: Record<string, Counter> = {}
	const artifactExamples: string[] = []
	let noPoly = 0

	for (const r of rows) {
		const st = r.state || "??"
		inc(overall, "n")
		byState[st] ??= {}
		inc(byState[st], "n")
		const nameOk = Boolean(r.nameMatch)

		if (nameOk) {
			inc(overall, "name")
			inc(byState[st]!, "name")
		}

		const lid = r.neuralLocID
		const contained = lid ? geometryContains(await geomForID(lid), r.lon, r.lat) : null

		if (contained !== null) {
			// a polygon existed and was tested (True or False)
			inc(overall, "poly")
			inc(byState[st]!, "poly")
		} else if (lid) {
			noPoly += 1
		}

		if (contained) {
			inc(overall, "pip")
			inc(byState[st]!, "pip")

			if (!nameOk && artifactExamples.length < MAX_LISTED_ARTIFACTS) {
				artifactExamples.push(`  "${r.input}"  gold="${pyStr(r.expectedLoc)}"  resolved="${pyStr(r.neuralLoc)}"`)
			}
		}
	}

	console.log(`\n=== PIP-containment vs name-match (${src}${labelArg ? " · " + labelArg : ""}) ===`)
	console.log(line("OVERALL", overall))

	for (const st of Object.keys(byState).toSorted()) {
		console.log(line(st, byState[st]!))
	}

	console.log(`\n  rows resolved-but-polygon-missing: ${noPoly}`)
	console.log(`\nMETRIC-ARTIFACT cases (name-match FAILED but gold point IS inside the resolved locality):`)

	for (const e of artifactExamples) {
		console.log(e)
	}

	if (jsonOut) {
		const n = get(overall, "n")

		const summary = {
			label: labelArg,
			source: src,
			n,
			name_match: n ? get(overall, "name") / n : null,
			pip_all: n ? get(overall, "pip") / n : null,
			pip_poly: get(overall, "poly") ? get(overall, "pip") / get(overall, "poly") : null,
			poly_coverage: n ? get(overall, "poly") / n : null,
			no_polygon: noPoly,
		}

		await writeLocalJSONFile(summary, jsonOut)

		console.error(`\nwrote summary → ${jsonOut}`)
	}

	return 0
}

runIfScript(import.meta, main)
